src/Raven.Server/Rachis/Follower.cs

```diff
@@ -687,13 +698,31 @@ private void MaybeNotifyLeaderThatWeAreStillAlive(TransactionOperationContext co
                 midpointTerm = _engine.GetTermForKnownExisting(context, midpointIndex);
             }
 
-
-            while (minIndex < maxIndex)
+            while ((midpointTerm == negotiation.PrevLogTerm && midpointIndex == negotiation.PrevLogIndex) == false)
             {
                 _engine.Timeout.Defer(_connection.Source);
 
                 _engine.ValidateTerm(_term);
-                
+
+                if (midpointIndex == negotiation.PrevLogIndex && midpointTerm != negotiation.PrevLogTerm)
+                {
+                    // our appended entries has been diverged, same index with different terms.
+                    var msg = $"Our appended entries has been diverged, same index with different terms. " +
+                              $"My index/term {midpointIndex}/{midpointTerm}, while yours is {negotiation.PrevLogIndex}/{negotiation.PrevLogTerm}.";
```

Comment:
Here, make sure that you explain that we will just get the snapshot and that's it.

However, why not go further back in time to find the first matching index/term pair?

Reply:
If this happened we need to go to try the leader's last log of the previous term, which is unknown on the follower side and we don't have an index on the term.

I see a lot of pitfalls in going further back, so maybe create a minor performance issue for 4.1 for this?

Reply:
Why? Isn't this just considering this as also something that I don't have a match and keep going back using binary search?